src/state_manager/mod.rs

```diff
@@ -649,6 +652,49 @@ where
         self.call_raw(message, chain_rand, &ts)
     }
 
+    pub async fn apply_on_state_with_gas(
+        self: &Arc<Self>,
+        tipset: Option<Arc<Tipset>>,
+        msg: Message,
+    ) -> anyhow::Result<ApiInvocResult> {
+        let ts = tipset.unwrap_or_else(|| self.cs.heaviest_tipset());
+
+        // Handle state forks
+        // TODO(elmattic): https://github.com/ChainSafe/forest/issues/3733
+
+        let from_a = self.resolve_to_key_addr(&msg.from, &ts).await?;
+
+        // Pretend that the message is signed. This has an influence on the gas
+        // cost. We obviously can't generate a valid signature. Instead, we just
+        // fill the signature with zeros. The validity is not checked.
+        let mut chain_msg = match from_a.protocol() {
+            Protocol::Secp256k1 => ChainMessage::Signed(SignedMessage::new_unchecked(
+                msg.clone(),
+                Signature::new_secp256k1(vec![0; SECP_SIG_LEN]),
+            )),
+            Protocol::Delegated => ChainMessage::Signed(SignedMessage::new_unchecked(
+                msg.clone(),
+                // In Lotus, delegated signatures have the same length as SECP256k1.
+                // This may or may not change in the future.
+                Signature::new(SignatureType::Delegated, vec![0; SECP_SIG_LEN]),
+            )),
+            _ => ChainMessage::Unsigned(msg.clone()),
+        };
+
+        let (_invoc_res, apply_ret, duration) = self
+            .call_with_gas(&mut chain_msg, &[], Some(ts), VMTrace::Traced)
+            .await?;
+        Ok(ApiInvocResult {
+            msg: msg.clone(),
```

Comment:
Is there a way to limit the number of clones for `msg` in this method?

Reply:
Done.